gammapy/datasets/map.py

```diff
@@ -2510,13 +2509,31 @@ def update(self, exposure, psf, edisp, geom):
         # TODO: simplify and clean up
         log.debug("Updating model evaluator")
         # cache current position of the model component
+        geom2d = geom.to_image()
```

Comment:
Maybe we should rely on the IRF geom here and not the counts geom? Maybe just pass directly `MapDataset.mask_safe_psf` or `MapDataset.mask_safe_edisp` to `.update()`? The coarser bin size will make the update faster and accessing the IRF mask ensures consistency...

Reply:
ok added mask_safe_psf to MapEvaluator argument, and some extra code to deal with the cases where mask_safe is None or psf is None.